[WIP] Adding smoke test for Spring for Apache Pulsar

onobc opened the pull request:
For commenting only

wilkinsona committed:
Add Gradle task to create a status page based on smoke test projects

wilkinsona committed:
Add Gradle task to update smoke-tests.yml from smoke test projects

sobychacko committed:
Adding smoke test for Spring for Apache Pulsar

onobc commented on integration/spring-pulsar/build.gradle:
```diff
@@ -0,0 +1,26 @@
+plugins {
+    id 'java'
+    id 'org.springframework.boot'
+    id 'org.springframework.boot.aot'
```
I don't think you need this explicitly.

onobc commented on integration/spring-pulsar/build.gradle:
```diff
@@ -0,0 +1,26 @@
+
+repositories {
+    mavenLocal()
+    maven {
```
It may be helpful to add a comment as to why this repo is here (backing reactive client).

onobc commented on integration/spring-pulsar/build.gradle:
```diff
@@ -0,0 +1,26 @@
+}
+
+repositories {
+    mavenLocal()
```
Instead of setting `mavenLocal()` take a look at the [`fromMavenLocal`](https://github.com/spring-projects/spring-aot-smoke-tests/blob/90e784470f6f0a4b2e328df3ea1fda415fea6bf5/gradle/plugins/aot-smoke-test-plugin/src/main/java/org/springframework/aot/gradle/AotSmokeTestPlugin.java#L81) property.

onobc commented on integration/spring-pulsar/build.gradle:
```diff
@@ -0,0 +1,26 @@
+
+dependencies {
+    implementation(platform(org.springframework.boot.gradle.plugin.SpringBootPlugin.BOM_COORDINATES))
+    implementation("org.springframework.boot:spring-boot-starter")
```
This is not needed as its included by the `spring-pulsar-spring-boot-starter` now.

onobc commented on integration/spring-pulsar/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+version: '3'
+services:
+  standalone:
```
We should rename this to `pulsar`.

onobc commented on integration/spring-pulsar/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+      - 8080
+      - 6650
+    command: bin/pulsar standalone
+    volumes:
```
Do we need volumes for the smoke test? I think we can ditch them altogether.

onobc commented on integration/spring-pulsar/src/main/java/com/example/pulsar/SpringPulsarApplication.java:
```diff
@@ -0,0 +1,42 @@
+
+    record Greeting(String message) {
+        @Override
+        public String toString() {
```
We can ditch this custom `toString` and just make the test look for that output. It is unlikely to change often.

onobc commented on integration/spring-pulsar/src/main/java/com/example/pulsar/SpringPulsarApplication.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Bean
+    ApplicationRunner runner1(PulsarTemplate<Greeting> pulsarTemplate) {
+        String topic = "graal-topic";
```
**[TinyNit]** I think it helps w/ consistency if we make the topic and subscription name prefixes match such as `graalvm-demo-topic`.

onobc commented on integration/spring-pulsar/src/main/java/com/example/pulsar/SpringPulsarApplication.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Bean
+    ApplicationRunner runner1(PulsarTemplate<Greeting> pulsarTemplate) {
```
**[TinyNit]** It may help w/ readability if we name the method something more descriptive along the lines of what it does, maybe something like `sendMessageToTopicOnAppStartup()`.

onobc commented on integration/spring-pulsar/src/main/java/com/example/pulsar/SpringPulsarApplication.java:
```diff
@@ -0,0 +1,42 @@
+
+    @PulsarListener(subscriptionName = "graalvm-demo-subscription", topics = "graal-topic",
+            schemaType = SchemaType.JSON)
+    void process(Greeting message) {
```
**[TinyNit]** It may help w/ readability if we name the method something more descriptive along the lines of what it does, maybe something like `receiveMessageFromTopic()`. It seems nit-picky but if someone looks in here to debug something later and they know nothing about spring-pulsar, this could make it easier to follow.

onobc commented on integration/spring-pulsar/src/main/java/com/example/pulsar/SpringPulsarApplication.java:
```diff
@@ -0,0 +1,42 @@
+        String topic = "graal-topic";
+        pulsarTemplate.setSchema(Schema.JSON(Greeting.class));
+        return args -> {
+            pulsarTemplate.sendAsync(topic, new Greeting("Hello from GraalVM!"));
```
It may be "safer" to use blocking send here. Ideally it does not matter because the awaitility of the tests will wait for it to exit etc..

onobc reviewed (commented):
@sobychacko looks great! Nice work!

**Also note** you will need to add to the `ci/smoke-test.yml` file.

sobychacko commented:
@onobc All valid and great comment. I think it might be ok with the async send, though.

onobc commented on integration/spring-pulsar/build.gradle:
```diff
@@ -0,0 +1,24 @@
+dependencies {
+    implementation(platform(org.springframework.boot.gradle.plugin.SpringBootPlugin.BOM_COORDINATES))
+    implementation("org.springframework.pulsar:spring-pulsar-spring-boot-starter:0.1.0-SNAPSHOT")
+    implementation "org.apache.pulsar:pulsar-client-reactive-adapter:0.1.0-SNAPSHOT"
```
If we are bringing in the reactive stuff we can replace these w/ the reactive pulsar starter as it includes both.

onobc replied:
That being said, is the plan to have a follow on PR that includes reactive support? If we do that then let's just include the imperative starter for now and remove this dep and also the repository from the build.gradle. Wdyt?

sobychacko replied:
I needed the reactive dependency because otherwise, the `nativeCompile` wasn't working.

onobc reviewed (commented):
@sobychacko looks good. couple of comments to address.

sobychacko commented:
This is happening without the reactor client dep.
```
Exception in thread "main" java.lang.NoClassDefFoundError: reactor/core/publisher/Flux
        at org.springframework.pulsar.aot.PulsarRuntimeHints.registerHints(PulsarRuntimeHints.java:61)
        at org.springframework.context.aot.RuntimeHintsBeanFactoryInitializationAotProcessor$RuntimeHintsRegistrarContribution.lambda$applyTo$0(RuntimeHintsBeanFactoryInitializationAotProcessor.java:116)
        at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:647)
        at org.springframework.context.aot.RuntimeHintsBeanFactoryInitializationAotProcessor$RuntimeHintsRegistrarContribution.applyTo(RuntimeHintsBeanFactoryInitializationAotProcessor.java:110)
        at org.springframework.context.aot.BeanFactoryInitializationAotContributions.applyTo(BeanFactoryInitializationAotContributions.java:78)
        at org.springframework.context.aot.ApplicationContextAotGenerator.lambda$processAheadOfTime$0(ApplicationContextAotGenerator.java:58)
        at org.springframework.context.aot.ApplicationContextAotGenerator.withCglibClassHandler(ApplicationContextAotGenerator.java:67)
        at org.springframework.context.aot.ApplicationContextAotGenerator.processAheadOfTime(ApplicationContextAotGenerator.java:53)
        at org.springframework.context.aot.ContextAotProcessor.performAotProcessing(ContextAotProcessor.java:106)
        at org.springframework.context.aot.ContextAotProcessor.doProcess(ContextAotProcessor.java:84)
        at org.springframework.context.aot.ContextAotProcessor.doProcess(ContextAotProcessor.java:49)
        at org.springframework.context.aot.AbstractAotProcessor.process(AbstractAotProcessor.java:82)
        at org.springframework.boot.SpringApplicationAotProcessor.main(SpringApplicationAotProcessor.java:76)
Caused by: java.lang.ClassNotFoundException: reactor.core.publisher.Flux
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
```

sobychacko committed:
Addressing PR review comments

mhalbritter committed:
Upgrade to NBT 0.9.18

onobc committed:
Add docker compose tips to README.adoc

mhalbritter committed:
Merge pull request #154 from onobc/patch-1

Add docker compose tips to README.adoc

wilkinsona committed:
Add a lifecycle task to update the status page and CI pipeline

wilkinsona committed:
Trigger native tests and native app tests once per day

Closes gh-152

mhalbritter committed:
Add contribution guide

wilkinsona committed:
Trigger tests at times throughout the day

wilkinsona committed:
Don't create empty groups in CI pipeline

mhalbritter committed:
Move the unit tests from commandlinerunner to separate project

See gh-142

mhalbritter committed:
Polish CONTRIBUTING.adoc

mhalbritter committed:
Update infrastructure

mhalbritter committed:
Rename tcf in boot to boot-tcf to prevent conflicts

mhalbritter committed:
Rename test class

mhalbritter committed:
Polish RandomPortTests

mhalbritter committed:
Polish RandomPortTests

mhalbritter committed:
Add tests for composite health contributor

Closes gh-120

mhalbritter committed:
Update NBT to 0.9.18

mhalbritter committed:
Update Kotlin to 1.7.21

mhalbritter committed:
Add test for WebMVC with Jetty

See gh-50

wilkinsona committed:
Exclude tasks that format AOT-generated code

Disabling them isn't enough to avoid all the tasks upon which
they depend from being executed. This means that AOT processing
would occur (so that the source was generated) and was available for
formatting, even though that formatting had been disabled.

This commit updates the plugin to exclude the AOT source formatting
tasks instead of disabling them. This prevents everything upon which
they depend from being executed unnecessarily.

Closes gh-155

mhalbritter committed:
Add test for WebFlux with Jetty

Closes gh-50

mhalbritter committed:
Use default reachability metadata version

wilkinsona committed:
Increase application start up timeout

Some infrastruture, such as Cassandra, doesn't become operational
quickly enough for an app to be able to start within 10 seconds.
This leads to intermittent build failures in our CI environment.

This commit increases the timeout period to 30 seconds.

mhalbritter committed:
Extract kotlin version into property

mhalbritter committed:
Extract java-format and NBT version into properties

mhalbritter committed:
Remove constraints from session-jdbc smoke test

mhalbritter committed:
Extract Spring Boot version into a property

mhalbritter committed:
Add data-jdbc-h2-kotlin smoke test

mhalbritter committed:
Test against Spring Boot 3.0.1-SNAPSHOT

mhalbritter committed:
Move @MockBean tests to boot-tcf

mhalbritter committed:
Format webmvc-tomcat

mhalbritter committed:
Add webmvc-undertow smoke test

See gh-58

mhalbritter committed:
Move 'websocket' to 'websocket-tomcat'

mhalbritter committed:
Add webflux-undertow smoke test

Closes gh-58

mhalbritter committed:
Increase cassandra timeouts

Closes gh-156

mhalbritter committed:
Remove useless Kotlin Gradle plugin from webmvc-tomcat-tls

mhalbritter committed:
Add webmvc-jetty-tls smoke test

See gh-68

mhalbritter committed:
Add webmvc-undertow-tls smoke test

Closes gh-68

mhalbritter committed:
Add logging-logback-xml smoke test

Closes gh-126

sobychacko committed:
Formatting changes

sobychacko committed:
Update infrastructure for Spring Pulsar

sobychacko committed:
Addressing PR review comments
